### PR TITLE
chore!: add native blob tx

### PR DESCRIPTION
**Pull request by cmwaters**
Closes: https://github.com/celestiaorg/go-square/issues/77, https://github.com/celestiaorg/go-square/issues/33

### CODE REVIEW

File: share/blob_tx.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	ProtoBlobTxTypeID = "BLOB"
+)
+
+type BlobTx struct {
+	Tx    []byte
+	Blobs []*Blob
+}
+
+// UnmarshalBlobTx attempts to unmarshal a transaction into blob transaction. If an
+// error is thrown, false is returned.
```
Comment (rootulp) [inline share/blob_tx.go:22]: IMO the second sentence is a bit misleading b/c this function can return an error and isBlob=false

```suggestion
// UnmarshalBlobTx attempts to unmarshal a transaction into blob transaction.
```

File: share/blob_tx.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+func UnmarshalBlobTx(tx []byte) (*BlobTx, bool, error) {
```
Comment (rootulp) [inline share/blob_tx.go:23]: [nit] I think providing a name for the second return parameter would help communicate it's meaning

```suggestion
func UnmarshalBlobTx(tx []byte) (blobTx *BlobTx, isBlobTx bool, err error) {
```

File: share/utils.go [hunk 3]
```diff
@@ -40,7 +40,7 @@ func zeroPadIfNecessary(share []byte, width int) (padded []byte, bytesOfPadding
 // in compact shares where units (i.e. a transaction) are prefixed with a length
 // delimiter that is encoded as a varint. Input should not contain the namespace
 // ID or info byte of a share.
-func ParseDelimiter(input []byte) (inputWithoutLenDelimiter []byte, unitLen uint64, err error) {
+func parseDelimiter(input []byte) (inputWithoutLenDelimiter []byte, unitLen uint64, err error) {
```
Comment (rootulp) [inline share/utils.go:43]: [nit]

go doc `ParseDelimiter` => `parseDelimiter`

**Review by rootulp — APPROVED** [review]
I'm a fan of distinct files for blob_tx.go and index_wrapper.go. Overall I'm a fan of one file per significant struct (i.e. `type BlobTx struct`)

File: share/blob_tx.go (hunk 1 above)
Comment (cmwaters) [inline share/blob_tx.go:22]:
```suggestion
// UnmarshalBlobTx attempts to unmarshal a transaction into blob transaction. It returns a boolean 
// If the bytes are of type BlobTx and an error if there is a problem with decoding
```

File: share/blob_tx.go (hunk 2 above)
Comment (cmwaters) [reply share/blob_tx.go:23]: I'm just going to try make it clear in the comments

**Review by rootulp — APPROVED** [review]

File: builder.go [hunk 4]
```diff
@@ -42,14 +43,17 @@ func NewBuilder(maxSquareSize int, subtreeRootThreshold int, txs ...[]byte) (*Bu
...
+		Pfbs:                 make([]*v1.IndexWrapper, 0),
 		Txs:                  make([][]byte, 0),
 		TxCounter:            share.NewCompactShareCounter(),
 		PfbCounter:           share.NewCompactShareCounter(),
 	}
 	seenFirstBlobTx := false
 	for idx, tx := range txs {
-		blobTx, isBlobTx := share.UnmarshalBlobTx(tx)
+		blobTx, isBlobTx, err := share.UnmarshalBlobTx(tx)
+		if err != nil && isBlobTx {
```
Comment (rach-id) [inline builder.go:54]: what if `err !=nil` but `isBlobTx == false`, where does the unmarshalling error get returned?

File: share/blob_tx.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	// ProtoBlobTxTypeID is included in each encoded BlobTx to help prevent
+	// decoding binaries that are not actually BlobTxs.
+	ProtoBlobTxTypeID = "BLOB"
+)
+
+type BlobTx struct {
```
Comment (Wondertan) [inline share/blob_tx.go:16]: Is the idea of making go-square hold actual transactions? Isn't that a concern of the app?

**Review by evan-forbes — APPROVED** [review]

**Pull request merged**